### PR TITLE
Doc for implementing an algorithm using tf and np.

### LINKED ISSUE #1671: Docs page "Tutorials > Implement a custom Algorithm (TF and Numpy sections)"

### PR DESCRIPTION
**Pull request by yeukfu**
This should close #1671.

### CODE REVIEW

File: docs/user/implement_algo.rst [hunk 1]
```diff
@@ -348,9 +358,107 @@ For completeness, the full experiment file is repeated below:
...
+.. code-block:: python
+
+    import tensorflow as tf
+    import numpy as np
+
+    from garage import log_performance, TrajectoryBatch
+    from garage.sampler import RaySampler
+    from garage.misc import tensor_utils
+
+    class MyAlgorithm():
```
Comment (ryanjulian) [inline docs/user/implement_algo.rst:376]: no trailing () needed

File: docs/user/implement_algo.rst [hunk 2]
```diff
@@ -212,10 +212,18 @@ generators used for the experiment.
     debug_my_algorithm()
```
Comment (ryanjulian) [inline docs/user/implement_algo.rst:212]: can you convert this docs page to markdown?

File: docs/user/implement_algo.rst [hunk 3]
```diff
@@ -348,9 +358,107 @@ For completeness, the full experiment file is repeated below:
...
+        def train(self, runner):
+            for epoch in runner.step_epochs():
+                samples = runner.obtain_samples(epoch)
+                log_performance(epoch,
+                                TrajectoryBatch.from_trajectory_list(self.env_spec,
+                                                                     samples),
+                                self._discount)
+                self.train_once(samples)
+
+        def train_once(self, paths):
```
Comment (ryanjulian) [inline docs/user/implement_algo.rst:413]: please make this private, since it's not part of the Algorithm API

File: docs/user/implement_algo.rst [hunk 4]
```diff
@@ -348,9 +358,107 @@ For completeness, the full experiment file is repeated below:
...
+            env = GarageEnv(PointEnv())
+            policy = GaussianMLPPolicy(env.spec)
+            algo = MyAlgorithm(env.spec, policy)
+            runner.setup(algo, env)
+            runner.train(n_epochs=500, batch_size=4000)
+
+    debug_my_algorithm()
+
 
 .. [1] Williams, Ronald J. "Simple statistical gradient-following algorithms for connectionist reinforcement learning." Machine learning 8.3-4 (1992): 229-256.
```
Comment (ryanjulian) [inline docs/user/implement_algo.rst:459]: please use bibtex for generating these citations. See https://github.com/rlworkgroup/garage/pull/1765 for an example.

File: docs/user/implement_algo.rst [hunk 5]
```diff
@@ -348,9 +358,107 @@ For completeness, the full experiment file is repeated below:
...
+
+.. code-block:: python
+
+    import tensorflow as tf
+    import numpy as np
+
+    from garage import log_performance, TrajectoryBatch
+    from garage.sampler import RaySampler
+    from garage.misc import tensor_utils
+
```
Comment (ryanjulian) [inline docs/user/implement_algo.rst:375]: please make sure this is formatted in a way that would pass pre-commit and is actually tested.

an easy way to do this is to actually make this a file in `examples/` and then directly include it here. 

there's an example of how to do this in https://github.com/rlworkgroup/garage/pull/1765

File: docs/user/implement_algo.rst [hunk 6]
```diff
@@ -348,9 +358,107 @@ For completeness, the full experiment file is repeated below:
...
+            del data['_returns']
+            return data
+
+    from garage import wrap_experiment
+    from garage.envs import PointEnv, GarageEnv
+    from garage.experiment import LocalTFRunner
+    from garage.experiment.deterministic import set_seed
+    from garage.tf.policies import GaussianMLPPolicy
+
+    @wrap_experiment(log_dir='my_algorithm_logs', use_existing_dir=True, archive_launch_repo=False)
```
Comment (ryanjulian) [inline docs/user/implement_algo.rst:446]: do we need these wrap_experiment args? i think they clutter the example

File: docs/user/implement_algo.rst (hunk 1 above)
Comment (ryanjulian) [inline docs/user/implement_algo.rst:376]: what does this algorithm actually do? why not just give it the real name?

File: docs/user/implement_algo.rst [hunk 7]
```diff
@@ -348,9 +358,107 @@ For completeness, the full experiment file is repeated below:
...
+            with tf.name_scope('loss'):
+                ll = policy_dist.log_prob(self._action, name='log_likelihood')
+                loss = -tf.reduce_mean(ll * self._returns)
+            with tf.name_scope('train'):
+                self._train_op = tf.compat.v1.train.AdamOptimizer(
+                    self._lr).minimize(loss)
+
+        def train(self, runner):
+            for epoch in runner.step_epochs():
+                samples = runner.obtain_samples(epoch)
```
Comment (ryanjulian) [inline docs/user/implement_algo.rst:406]: can you make this example operate on `TrajectoryBatch` directly?

File: docs/user/implement_algo.rst [hunk 8]
```diff
@@ -212,10 +212,18 @@ generators used for the experiment.
...
 Of course, we'll need to actually use the resulting samples to train our
-policy. For the purposes of this tutorial, we'll implement an extremely simple
-form of REINFORCE [1]_ (a.k.a. Vanilla Policy Gradient) using PyTorch.
+policy with PyTorch, TensorFlow or NumPy. For the purposes of this tutorial,
+we'll implement an extremely simple form of REINFORCE [1]_ (a.k.a. Vanilla
+Policy Gradient) using PyTorch and TensorFlow. If you want to use NumPy, you
+can refer `cem.py <https://github.com/rlworkgroup/garage/blob/master/src/garage/np/algos/cem.py>`_,
+which implements a simple Cross Entropy Method algorithm.
 
+PyTorch
```
Comment (ryanjulian) [inline docs/user/implement_algo.rst:225]: I think in both cases you should include console output and even plots for what the user should expect to see. Otherwise, they won't have a way of knowing if their example is working.

File: docs/user/implement_algo.rst [hunk 9]
```diff
@@ -212,10 +212,18 @@ generators used for the experiment.
...
+Trainning the Policy with Samples
+=================================
+
 Of course, we'll need to actually use the resulting samples to train our
-policy. For the purposes of this tutorial, we'll implement an extremely simple
-form of REINFORCE [1]_ (a.k.a. Vanilla Policy Gradient) using PyTorch.
+policy with PyTorch, TensorFlow or NumPy. For the purposes of this tutorial,
+we'll implement an extremely simple form of REINFORCE [1]_ (a.k.a. Vanilla
+Policy Gradient) using PyTorch and TensorFlow. If you want to use NumPy, you
+can refer `cem.py <https://github.com/rlworkgroup/garage/blob/master/src/garage/np/algos/cem.py>`_,
```
Comment (ryanjulian) [inline docs/user/implement_algo.rst:222]: Please document implementing numpy algorithms explicitly, even if it's very different than TF/Torch

File: docs/user/implement_algo.rst (hunk 1 above)
Comment (krzentner) [inline docs/user/implement_algo.rst:376]: You don't need `()` here.

File: docs/user/implement_algo.rst [hunk 10]
```diff
@@ -348,9 +358,107 @@ For completeness, the full experiment file is repeated below:
...
+            self.env_spec = env_spec
+            self.policy = policy
+            self.max_path_length = 200
+            self._lr = 1e-3
+            self._discount = 0.99
+
+            obs_space = self.policy.observation_space
+            action_space = self.policy.action_space
+            with tf.name_scope('inputs'):
+                self._obs = obs_space.to_tf_placeholder(name='obs', batch_dims=2)
```
Comment (krzentner) [inline docs/user/implement_algo.rst:390]: Would this file be simpler if we only used one batch dimension and we concatenated (instead of padding and stacking) the trajectories? I don't have a very strong opinion, but we should try and keep this example relatively simple.

File: docs/user/implement_algo.rst [hunk 11]
```diff
@@ -348,9 +358,107 @@ For completeness, the full experiment file is repeated below:
...
+            sess = tf.compat.v1.get_default_session()
+            sess.run(self._train_op, feed_dict={
+                 self._obs: np.stack(obs, axis=0),
+                 self._action: np.stack(actions, axis=0),
+                 self._returns: np.stack(returns, axis=0),
+            })
+            return np.mean(returns)
+
+        # For pickling
+        def __getstate__(self):
```
Comment (krzentner) [inline docs/user/implement_algo.rst:432]: You should probably label this "For pickling / snapshot" and point to the other docs which are relevant (probably in the paragraph above this source block). We should really also have `__setstate__` here (probably using a method named `_init_opt` or something), to make the whole picture clear.

File: docs/user/implement_algo.rst (hunk 3 above)
Comment (krzentner) [inline docs/user/implement_algo.rst:413]: Can you actually rename `train_once` to `_train_once` throughout this file to make it clear it's not part of the API?

File: docs/user/implement_algo.rst (hunk 9 above)
Comment (krzentner) [inline docs/user/implement_algo.rst:222]: Should be "can refer to cem.py"

File: docs/user/implement_algo.rst [hunk 12]
```diff
@@ -322,7 +331,8 @@ For completeness, the full experiment file is repeated below:
                 returns = torch.Tensor(returns_numpy.copy())
                 obs = torch.Tensor(path['observations'])
                 actions = torch.Tensor(path['actions'])
-                log_likelihoods = self.policy.log_likelihood(obs, actions)
+                dist = self.policy(obs)[0]
+                log_likelihoods = dist.log_prob(actions)
```
Comment (krzentner) [inline docs/user/implement_algo.rst:335]: Thanks for fixing this!

File: docs/user/implement_algo.rst (hunk 7 above)
Comment (yeukfu) [reply docs/user/implement_algo.rst:406]: Will this make it complex? As in the PyTorch version, KR use `TrajectoryBatch` only for logging.

**Comment by yeukfu** [comment]
@ryanjulian @krzentner It is ready for review.

File: docs/user/implement_algo.md [hunk 13]
```diff
@@ -0,0 +1,635 @@
...
+class RLAlgorithm(abc.ABC):
+    """Base class for all the algorithms.
+
+    Note:
+        If the field sampler_cls exists, it will be by LocalRunner.setup to
+        initialize a sampler.
+
+    """
+
+    # pylint: disable=too-few-public-methods
```
Comment (ryanjulian) [inline docs/user/implement_algo.md:37]: please remove

File: examples/np/tutorial_cem.py [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+    set_seed(100)
+    with LocalTFRunner(ctxt) as runner:
+        env = GarageEnv(env_name='CartPole-v1')
+        policy = CategoricalMLPPolicy(env.spec)
+        algo = SimpleCEM(env.spec, policy)
+        runner.setup(algo, env)
+        runner.train(n_epochs=100, batch_size=1000, plot=True)
+
+
+debug_my_algorithm()
```
Comment (ryanjulian) [inline examples/np/tutorial_cem.py:124]: can you rename these experiments to match the file? e.g. `tutorial_cem` and `tutorial_vpg`?

File: docs/user/implement_algo.md [hunk 15]
```diff
@@ -0,0 +1,635 @@
...
+
+The interface requires a single method, `train(runner)`, which takes a
+`garage.experiment.LocalRunner`. The interface is defined in
+`garage.np.algos.RLAlgorithm`, but inheriting from this class isn't necessary.
+
+Some additional functionality (such as sampling and plotting) require additional
+fields to exist.
+
+```py
+class RLAlgorithm(abc.ABC):
```
Comment (ryanjulian) [inline docs/user/implement_algo.md:28]: can you use a literal include for this? 

see https://github.com/rlworkgroup/garage/pull/1765/files#diff-f6fdd52adff0a248b20cbb76973f87e0R42 for an example

File: docs/user/implement_algo.md [hunk 16]
```diff
@@ -0,0 +1,635 @@
...
+Of course, we'll need to actually use the resulting samples to train our
+policy with PyTorch, TensorFlow or NumPy. In this tutorial, we'll implement an
+extremely simple form of REINFORCE :cite:`williams1992simple` (a.k.a. Vanilla
+Policy Gradient) using PyTorch and TensorFlow. We will also implement a simple
+Cross Entropy Method (CEM) using NumPy.
+```
+
+### PyTorch
+
+```py
```
Comment (ryanjulian) [inline docs/user/implement_algo.md:212]: can this use a literal include?

File: docs/user/implement_algo.md [hunk 17]
```diff
@@ -0,0 +1,635 @@
...
+        runner.train(n_epochs=500, batch_size=4000)
+...
+```
+
+Different from PyTorch's version, we need to build the computation graph before
+training the policy in TensorFlow.
+
+```py
+...
+import tensorflow as tf
```
Comment (ryanjulian) [inline docs/user/implement_algo.md:369]: can this use a literal include?

File: docs/user/implement_algo.md [hunk 18]
```diff
@@ -0,0 +1,635 @@
...
+
+### NumPy
+
+We will implement [CEM](https://github.com/rlworkgroup/garage/blob/master/src/garage/np/algos/cem.py)
+with NumPy, and train the `CategoricalMLPPolicy` to solve `CartPole-v1`. The
+debug function is similar to that of TensorFlow:
+
+```py
+@wrap_experiment
+def debug_my_algorithm(ctxt=None):
```
Comment (ryanjulian) [inline docs/user/implement_algo.md:512]: please choose better experiment names for these examples

File: docs/user/implement_algo.md [hunk 19]
```diff
@@ -0,0 +1,635 @@
...
+```
+
+## Training the Policy with Samples
+
+```eval_rst
+Of course, we'll need to actually use the resulting samples to train our
+policy with PyTorch, TensorFlow or NumPy. In this tutorial, we'll implement an
+extremely simple form of REINFORCE :cite:`williams1992simple` (a.k.a. Vanilla
+Policy Gradient) using PyTorch and TensorFlow. We will also implement a simple
+Cross Entropy Method (CEM) using NumPy.
```
Comment (ryanjulian) [inline docs/user/implement_algo.md:207]: please cite CEM https://scholar.google.com/scholar?cluster=17794203204847200474&hl=en&as_sdt=2005&sciodt=0,5

File: docs/user/implement_algo.md [hunk 20]
```diff
@@ -0,0 +1,635 @@
...
+----------------------------------  ---------
+```
+
+### NumPy
+
+We will implement [CEM](https://github.com/rlworkgroup/garage/blob/master/src/garage/np/algos/cem.py)
+with NumPy, and train the `CategoricalMLPPolicy` to solve `CartPole-v1`. The
+debug function is similar to that of TensorFlow:
+
+```py
```
Comment (ryanjulian) [inline docs/user/implement_algo.md:510]: can this use a literal include of your example file?

File: docs/user/implement_algo.md [hunk 21]
```diff
@@ -0,0 +1,635 @@
...
+Evaluation/MaxReturn                  18
+Evaluation/MinReturn                  10
+Evaluation/NumTrajs                  299
+Evaluation/StdReturn                   1.81062
+Evaluation/TerminationRate             1
+TotalEnvSteps                       8015
+----------------------------------  ----------
+```
+
+## Reference
```
Comment (ryanjulian) [inline docs/user/implement_algo.md:625]: References (plural)

File: docs/user/implement_algo.md [hunk 22]
```diff
@@ -0,0 +1,434 @@
...
+extremely simple form of REINFORCE :cite:`williams1992simple` (a.k.a. Vanilla
+Policy Gradient) using PyTorch and TensorFlow. We will also implement a simple
+Cross Entropy Method (CEM) :cite:`rubinstein2004cross` using NumPy.
+```
+
+### PyTorch
+
+```eval_rst
+.. literalinclude:: ../../examples/torch/tutorial_vpg.py
+   :lines: 3-5,10-11,16,24-31,33,40-41,46-48,58-72
```
Comment (ryanjulian) [inline docs/user/implement_algo.md:195]: i'm pretty concerned that using the line numbers here, because if these files get changes, this documentation might silently look really wrong. does it work if you include only whole files?

Comment (yeukfu) [reply docs/user/implement_algo.md:195]: It sure works if whole files are included. But in the tutorial, we need to split it into several parts to make it clearer and focusing.

Comment (yeukfu) [reply docs/user/implement_algo.md:195]: So the other choice is copy and paste code blocks except for the final complete file.

**Comment by yeukfu** [comment]
@Mergifyio refresh

**Comment by yeukfu** [comment]
@Mergifyio rebase

**Comment by ryanjulian** [comment]
@Mergifyio rebase
